code/datums/brain_damage/imaginary_friend.dm

```diff
@@ -181,53 +183,242 @@
+		SSblackbox.record_feedback("tally", "ic_blocked_words", 1, lowertext(config.ic_filter_regex.match))
+		return
+
+	if(soft_filter_result && !filterproof)
+		if(tgui_alert(usr,"Your message contains \"[soft_filter_result[CHAT_FILTER_INDEX_WORD]]\". \"[soft_filter_result[CHAT_FILTER_INDEX_REASON]]\", Are you sure you want to say it?", "Soft Blocked Word", list("Yes", "No")) != "Yes")
+			SSblackbox.record_feedback("tally", "soft_ic_blocked_words", 1, lowertext(config.soft_ic_filter_regex.match))
+			log_filter("Soft IC", message, filter_result)
 			return
+		message_admins("[ADMIN_LOOKUPFLW(usr)] has passed the soft filter for \"[soft_filter_result[CHAT_FILTER_INDEX_WORD]]\" they may be using a disallowed term. Message: \"[message]\"")
+		log_admin_private("[key_name(usr)] has passed the soft filter for \"[soft_filter_result[CHAT_FILTER_INDEX_WORD]]\" they may be using a disallowed term. Message: \"[message]\"")
+		SSblackbox.record_feedback("tally", "passed_soft_ic_blocked_words", 1, lowertext(config.soft_ic_filter_regex.match))
+		log_filter("Soft IC (Passed)", message, filter_result)
+
+	if(client && !(ignore_spam || forced))
+		if(client.prefs.muted & MUTE_IC)
+			to_chat(src, span_danger("You cannot speak IC (muted)."))
+			return FALSE
+		if(client.handle_spam_prevention(message, MUTE_IC))
+			return FALSE
 
-	friend_talk(message)
+	friend_talk(message, spans, forced)
 
 /mob/camera/imaginary_friend/Hear(message, atom/movable/speaker, datum/language/message_language, raw_message, radio_freq, list/spans, list/message_mods = list())
 	if (client?.prefs.read_preference(/datum/preference/toggle/enable_runechat) && (client.prefs.read_preference(/datum/preference/toggle/enable_runechat_non_mobs) || ismob(speaker)))
 		create_chat_message(speaker, message_language, raw_message, spans)
 	to_chat(src, compose_message(speaker, message_language, raw_message, radio_freq, spans, message_mods))
 
-/mob/camera/imaginary_friend/proc/friend_talk(message)
-	message = capitalize(trim(copytext_char(sanitize(message), 1, MAX_MESSAGE_LEN)))
+/mob/camera/imaginary_friend/proc/friend_talk(message, list/spans, forced = null)
+	var/list/message_mods = list()
+	message = get_message_mods(message, message_mods)
+	message = capitalize(message)
 
-	if(!message)
+	if(message_mods[RADIO_EXTENSION] == MODE_ADMIN)
+		client?.cmd_admin_say(message)
 		return
 
-	src.log_talk(message, LOG_SAY, tag="imaginary friend")
+	if(message_mods[RADIO_EXTENSION] == MODE_DEADMIN)
+		client?.dsay(message)
+		return
 
-	var/rendered = "<span class='game say'>[span_name("[name]")] <span class='message'>[say_quote(message)]</span></span>"
-	var/dead_rendered = "<span class='game say'>[span_name("[name] (Imaginary friend of [owner])")] <span class='message'>[say_quote(message)]</span></span>"
+	if(check_emote(message, forced))
+		return
 
-	to_chat(owner, "[rendered]")
-	to_chat(src, "[rendered]")
+	if(message_mods[MODE_SING])
+		var/randomnote = pick("\u2669", "\u266A", "\u266B")
```

Comment:
You should just put the characters in here directly instead, if our tools can't handle it we need to fix them

Reply:
Sure, will do. I tried doing the same thing living say does but I suppose that's for another PR